File: second_year/projects/G-OOP-400-COT-4-1-bsraytracer-13/src/camera.cpp
/*
** EPITECH PROJECT, 2026
** snd_year_projects
** File description:
** camera
*/

#include "../include/camera.hpp"
using namespace RayTracer;

Camera::Camera()
    : origin(Point3D(0, 0, 0)), screen(Point3D(-1, -1, -1), Vector3D(2, 0, 0), Vector3D(0, 2, 0))
{
}

Ray Camera::ray(double u, double v)
{
    Point3D _at = screen.pointAt(u, v);
    return Ray(origin, Vector3D(_at.X - origin.X, _at.Y - origin.Y, _at.Z - origin.Z));
}

File: second_year/projects/G-OOP-400-COT-4-1-bsraytracer-13/rust/src/camera.rs
//
// EPITECH PROJECT, 2026
// snd_year_projects
// File description:
// camera
//

use crate::Point3D;
use crate::Rectangle3D;
use crate::Ray;

pub struct Camera {
    pub origin: Point3D,
    pub screen: Rectangle3D,
}

impl Camera {
    pub fn ray(&self, u: f32, v: f32) -> Ray {
        return Ray { origin: self.origin, direction: self.screen.pointAt(u, v) - self.origin };
    }

    pub fn new(point: Point3D, rectangle: Rectangle3D) -> Self {
        return Camera { origin: point, screen: rectangle };
    }
}
